hubblestack/status.py

```diff
@@ -0,0 +1,413 @@
+        r['__doc__'] = {
+            'service.name.here': {
+                "count": 'number of times the counter was called',
+                "ema_dur": 'average duration of the calls',
+                "dt": 'time since the last call of the counter',
+                "ema_dt": 'average time between calls',
+                "dur": 'duration of the last call',
+                "last_t": 'the last time the counter was called',
+            },
+            'HEALTH': {
+                "last_activity": {
+                    "dt": 'the minimum dt across all tracked counters',
+                    "time": 'the time of the most recent counter',
+                },
+                "alive": {
+                        'yes': 'something was called within the last 60s',
+                        'warn': 'something was called within the last 300s',
+                        'hung': 'nothing has been called in 600s minutes',
+                        'unknown': 'unknown — probably not a good sign though',
+                },
+            },
+        }
+        h2['alive'] = 'unknown'
+        if h1['dt'] >= get_hubble_status_opt('hung_time'):
+            h2['alive'] = 'hung'
+        if h1['dt'] >= get_hubble_status_opt('warn_time'):
+            h2['alive'] = 'warn'
+        if h1['dt'] <= get_hubble_status_opt('good_time'):
+            h2['alive'] = 'yes'
+        return r
+
+    @classmethod
+    def as_json(cls, indent=2):
+        return json.dumps(cls.stats(), indent=indent)
+
+    @classmethod
+    def dumpster_fire(cls):
+        ''' dump the status.json file to cachedir
+
+            Location and filename can be adjusted with the cachedir and
+            hubble:status:dumpster options (see above).
+        '''
+        dumpster = get_hubble_status_opt('dumpster') or 'status.json'
+        if not dumpster.startswith('/'):
+            cachedir = get_hubble_or_salt_opt('cachedir') or '/tmp'
+            dumpster = os.path.join(cachedir, dumpster)
+        try:
+            with open(dumpster, 'w') as fh:
+                fh.write(cls.as_json())
+                fh.write('\n')
+            log.info("wrote HubbleStatus to %s", dumpster)
+        except:
+            log.exception("ignoring exception during dumpster fire")
+
+    @classmethod
+    def start_sigusr1_signal_handler(cls):
+        ''' start the signal.SIGUSR1 handler (dumps status to
+            /var/cache/hubble/status.json or whatever is specified in
+            cachedir + hubble:status:dumpster configs)
+        '''
+        if not cls._signaled:
+            cls._signaled = True
+            if not hasattr(signal, 'SIGUSR1'):
+                # TODO: invent invocation that works in windows instead of just complaining
+                log.info("signal package lacks SIGUSR1, skipping SIGUSR1 status.json handler setup")
+                return
+            signal.signal(signal.SIGUSR1, cls.dumpster_fire)
+
+def __setup_for_testing():
```

Comment:
Did you intend to have two leading underscores here? The only use-case for two leading underscores (without two trailing underscores) is for [name mangling](https://docs.python.org/3/tutorial/classes.html#private-variables). Can you either make this a true dunder (with two leading underscores) or just a private function (single leading underscore)?

Reply:
nah, I did that because I thought the double underscore had something to do with making the method super duper secret. I can change to a single underscore or simply remove the thing. I only use it for testing the status module outside of the running hubble and only by hand in ipython.

Note that I used the same goofy name in hubblestack.hec.opt -- possibly should remove that one too?

Reply:
Yeah probably. This is fine for now. Thanks!